Licenses controller

JoeCohen opened the pull request:
- Adds an admin-only UI with standard CRUD actions for Licenses.
- Updates and rewrites License portion of `/info/how_to_use` to include new licenses

#### Some Suggested Manual Tests
- go into Admin mode
- click License in the sidebar
- view various Licenses
- add a new License, then destroy it
- show an existing License. Note lack of a Destroy link.
- Edit some Licenses to: blank out an attribute; to duplicate another License's attribute.

JoeCohen committed:
First draft of Licenses#show

- Adds test, routes, controller, #show, and template
- Includes helpter, translation strings

JoeCohen committed:
Add skeleton tests for other actions

JoeCohen committed:
First draft of Licenses#index

JoeCohen committed:
Test showing non-existent license

JoeCohen committed:
Initial draft of Licenses#new

JoeCohen committed:
Add Licenses#create

JoeCohen committed:
Merge branch 'jdc-2151-licenses_controller' of https://github.com/MushroomObserver/mushroom-observer into jdc-2151-licenses_controller

JoeCohen committed:
Fix messed up merge

JoeCohen committed:
Add some validation

JoeCohen committed:
Add Licenses#edit

JoeCohen committed:
Add Licenses#update

JoeCohen committed:
Rubocopping

JoeCohen committed:
Fix typo in fixture used in test

- Fixes test failure

JoeCohen committed:
Sort Licenses in test to newest one

Avoids potential test bug

JoeCohen committed:
Add License fixture

JoeCohen committed:
Add License#destroy

JoeCohen committed:
Tweak nonexistent License in test

It's more convenient for me

JoeCohen committed:
Prevent destruction of License that's in use

JoeCohen committed:
Move helper methods inside of module.

How did this ever work?

JoeCohen committed:
Add Create License link to Index

JoeCohen committed:
Revise method name for consistency

JoeCohen committed:
Test show_license tabs (Red)

JoeCohen committed:
Move title & tab helpers to separate file

JoeCohen committed:
Add License#in_use? method

JoeCohen committed:
Add tabs to show_license page

JoeCohen committed:
Allow changing `deprecated` attribute

- Add Deprecated checkbox
- Save results of checking the box

JoeCohen committed:
Fix "successfully updated" flash

JoeCohen committed:
Add form tabs (link to Index)

JoeCohen committed:
Add Licenses link to admin sidebar

JoeCohen committed:
Add edit page tabs (links)

JoeCohen committed:
Merge branch 'jdc-2151-licenses_controller' of https://github.com/MushroomObserver/mushroom-observer into jdc-2151-licenses_controller

JoeCohen committed:
Add before_action callbacks

JoeCohen committed:
Update License section of How To Use

- Includes all the added licenses
- Updates descriptions. (Cribbed from https://www.inaturalist.org/posts/30692-tech-tip-tuesday-understanding-licensing#:~:text=However%2C%20what%20many%20don%E2%80%99t%20realize%20is%20that%20iNaturalist,categories%3A%20%E2%80%9Call%20rights%20reserved%E2%80%9D%20copyright%20and%20creative%20commons.)
- Fixes the Public Domain license link

JoeCohen committed:
Update license form links to https

JoeCohen committed:
Templates for badges for new and updated licenses

JoeCohen committed:
Update flashes for create, update, destroy

JoeCohen committed:
Fix create params level, misc changes

JoeCohen committed:
Fix `deprecated` values & changing to missing value

JoeCohen committed:
Rubocopping

JoeCohen committed:
Move `attributes_duplicated?` to model

JoeCohen committed:
Delete extra blank line

JoeCohen committed:
Remove unnecesary instance variable

JoeCohen committed:
Document License#attribute_duplicated?

JoeCohen commented on app/views/controllers/shared/_form_ccby40.erb:
```diff
@@ -0,0 +1,9 @@
+<%=
+tag.div(id: "license") do
+  link_to("https://creativecommons.org/licenses/by/4.0/", rel: "license") do
+    image_tag("https://i.creativecommons.org/l/by/4.0/88x31.png",
+              alt: "Creative Commons Attribution License",
+              class: "border-none")
+  end
+end
+%>
```
Template for license badge. 
This and the following templates anticipate adding these licenses to the db.

mo-nathan replied:
It seems odd to me that there are partials for each license if we are expecting admins to add licenses   Perhaps the license object can be parameterized with the CC.org links so there can be a single generic partial.

JoeCohen replied:
Thanks for the suggestion. I'll investigate way(s) of having a generic partial.

JoeCohen replied:
@mo-nathan This turns out to be very easy; CC has changed urls so that we don't even have to parametize the license object. In fact, I'm going to ditch the partial and put the code in the show template.

JoeCohen committed:
Fix filename of template for ccbyncnd badge

JoeCohen commented on test/integration/capybara/licenses_integration_test.rb:
```diff
@@ -0,0 +1,14 @@
+
+require("test_helper")
+
+# Test things that are untestable in integration tests
+class LicensesIntegrationTest < CapybaraIntegrationTestCase
+  def test_link_to_licenses
+    login(users(:admin))
+    assert_no_selector(:link, text: :LICENSES.l)
+
+    first("button", text: "Turn on Admin Mode").click
+    assert_selector(:link, text: :LICENSES.l)
+  end
+end
```
Can I do this in a functional test:
- Prove that when you're in admin mode, the sidebar has a `Licenses` link.

JoeCohen commented on app/models/license.rb:
```diff
@@ -45,15 +48,19 @@ class License < AbstractModel
   has_many :name_descriptions
```
@nwilson-eol , @pellaea:
### Should I add a License `created_at` column?
Seems weird that there isn't one, but that (obviously) doesn't break anything.

(PS: 3 of the Licenses have a updated_at == nil. That also seems weird. 
2 will be changed (shortly after this PR is deployed) when they are deprecated. The third (publicdomain) won't be changed.)

mo-nathan replied:
It does seem odd that there isn't a created_at.  I note that true of the nearby "interests" table as well.  Not sure what the thinking there was other than maybe trying to save some space?  At this point I'd rather it was all consistent with both a created_at and an updated_at.

JoeCohen replied:
Thanks! I will:
- Add a `created_at` column as a migration in this PR.
- Insure that `created_at` and `updated_at` are present for all Licenses by -- after this PR is merged and Licenses are updated via the new UI -- making any needed fixes on the back end.

JoeCohen committed:
Test LicensesHelper

- Covers newly added, but uncovered line

JoeCohen committed:
Less confusing method name

JoeCohen committed:
Increase LicenseController coverage to 100%

mo-nathan commented on app/views/controllers/licenses/index.html.erb:
```diff
@@ -0,0 +1,13 @@
+  flash_error(@error) if @error && @objects.empty?
+%>
+
+<%= paginate_block(@pages) do %>
```
Not a blocker, but I'd rather see this as a table than as what appears to be essentially a CSV.

mo-nathan replied:
I'd also include the "display_name" in the table since that's more readable than the "form_name".

mo-nathan commented:
The "form_name" field seems a bit odd from a naming perspective.  Putting "Form" in the show form doesn't make sense to me.  I'm guessing the "form_name" has to do with how it gets used either to generate URLs or maybe for the form submission?  Didn't dig in deeper, but it seems worth at least renaming to "short_name" or something along those lines.

mo-nathan commented:
Overall this seems like really good progress on getting this more easily manageable for admins.  Thanks for your work on it!

JoeCohen commented:
@nathan: Thanks for looking at this and for all your suggestions.

JoeCohen committed:
Add created_at to Licenses

JoeCohen committed:
Merge branch 'jdc-2151-licenses_controller' of https://github.com/MushroomObserver/mushroom-observer into jdc-2151-licenses_controller

JoeCohen committed:
Simplify code which displays License badges.

- Responds to [@mo-nathan's suggestion](https://github.com/MushroomObserver/mushroom-observer/pull/2154/files/a5819d6e1ff467f035d47ff4e329b4f8de0ae103#r1623392823)
- Add License#badge_url, derived from License#url
  (CC has made this possible by standardizing the badge urls. See https://licensebuttons.net/l/
- Use one partial for all license badges, deleting per license partials.

JoeCohen committed:
Adjust translation symbol name

JoeCohen committed:
Rename `form_name` => `code`

- Responds to @mo-nathan's suggestion. See https://github.com/MushroomObserver/mushroom-observer/pull/2154#issuecomment-2143816910

JoeCohen committed:
Delete Licenses `code` attriibute (fka `form_name`)

No longer needed, as all data is contained in (or can be derived from) License#url

JoeCohen committed:
Change index from "csv" style to table.

- Responds to [@mo-nathan's sugestions](https://github.com/MushroomObserver/mushroom-observer/pull/2154#discussion_r1623392075)

JoeCohen committed:
Delete obsole "code" translations

JoeCohen committed:
Show if License is default

JoeCohen committed:
Remove breakpoint -- oops

JoeCohen committed:
Revise License::preferred to get id from schema

JoeCohen committed:
Update License.rb documentation

JoeCohen committed:
100% coverage of License.rb

- Adds tests
- Fixes bug in `copyright_text`

JoeCohen committed:
100% coverage of LicensesController

JoeCohen committed:
Rubocop License.rb

JoeCohen committed:
Rubocop LicensesController

- Makes `create` and `update` ABC <= 17

JoeCohen committed:
Update .rubocop TODO file

- Just removes license fiels (doesn't update offense counts)

JoeCohen committed:
Fix unsafe html in License.rb

- Fixes failures introduced by #fd93eecbadee00b5627df40e20e964893f4964c7

JoeCohen committed:
Better name for method

JoeCohen committed:
Delete obsolete translation term
